### ISSUE TITLE
We need to clarify why we use median on EQ fatality calculations

### ISSUE THREAD

**Issue by timlinux** [inasafe/inasafe#3559]
### Problem

In https://github.com/inasafe/inasafe/blob/version-3_5_4/safe/impact_functions/earthquake/itb_earthquake_fatality_model/impact_function.py#L283 median is used on the fatalities array before fatalities are subtracted from exposed.

```python
            # Calculate expected number of displaced people per level
            displacements = displacement_rate[mmi] * (
                exposed - numpy.median(fatalities))

```

It would be good to clarify with a comment why we subtract the median of fatalities rather just the fatalities themselves.

CC @Charlotte-Morgan @wonder-sk

**Comment by Charlotte-Morgan** [comment]
@dynaryu @RikkiWeber - greatful if you can please review and discuss with EQ experts at GA to provide advice this week as we march toward InaSAFE 4.0 release

**Comment by wonder-sk** [comment]
Doing more inspection of the code, I have realized that the ITB "standard" model and Pager model assign a scalar value for displacement rate for each MMI level, however in ITB Bayesian model, the displacement rate for each MMI level is a vector of 15999 elements (the code reads the vectors from an auxiliary file worden_berngamma_log_fat_rate_inasafe_10.csv).

The variable "fatalities" is calculated as "number of exposed people" multiplied by "displacement rate", which is either a scalar or a vector.

Now in case of ITB Bayesian model, what is really unclear to me is that if we just always use median of the displacement rate vector (multiplication by number of exposed people does not change the median value), why not pre-calculate the median values from the CSV file into simple scalar values? That would make the code simpler and there would be no need to ship the 2.7mb extra CSV file...

**Comment by ismailsunni** [comment]
> Now in case of ITB Bayesian model, what is really unclear to me is that if we just always use median of the displacement rate vector (multiplication by number of exposed people does not change the median value), why not pre-calculate the median values from the CSV file into simple scalar values? That would make the code simpler and there would be no need to ship the 2.7mb extra CSV file...

+1

**Comment by dynaryu** [comment]
The ITB bayesian model provides a distribution of fatalities, instead of a scalar value. If the distribution is used, then the number of displaced also has a distribution.

**Comment by wonder-sk** [comment]
hi @dynaryu thanks for the clarifications!

The way how number of displaced people is calculated is indeed inconsistent with calculation of fatalities - see #3103.

Would it be possible to shed some more light on the associated data in the file ```worden_berngamma_log_fat_rate_inasafe_10.csv``` ? I understand that each column represents one MMI level and each level's fatality rate is modeled as a vector of ~16K probabilities, but the origin of the values is unclear to me. Is each row corresponding to a particular sub-model?

I have tried to find the referenced paper online, but with no luck unfortunately:
```
An Empirical Fatality Model for Indonesia Based on a Bayesian Approach
    by W. Sengara, M. Suarjana, M.A. Yulman, H. Ghasemi, and H. Ryu
    submitted for Journal of the Geological Society
```
I have found this paper "Development of Empirical Earthquake Fatality Model
For Indonesia" by Sengara et al (http://www.iitk.ac.in/nicee/wcee/article/WCEE2012_3641.pdf) but that one does not mention the Bayesian approach. Is there any chance to get hold of the original paper?

I guess the main question is whether to report the number of predicted fatalities as single number or whether to move to reporting of fatalities as a probability distribution like in USGS:

![image](https://cloud.githubusercontent.com/assets/193367/21916784/d2c3881a-d97d-11e6-9d75-1b3b98a51ad3.png)

InaSAFE currently displays just a single number of fatalities number, the magnitudes plot is only added to the metadata of output layer. To me it seems that reporting fatalities in the plot of magnitudes much better illustrates the uncertainty.

@Charlotte-Morgan @timlinux How should we proceed with this?

**Comment by Charlotte-Morgan** [comment]
hi @dynaryu  - can you please pass this request to Hadi if you prefer.

@RikkiWeber

**Comment by Gustry** [comment]
@wonder-sk made some testing in the PR below :

Testing results:
- Padang earthquake scenario
- extent 100.2077, -1.0384, 100.5444, -0.7960
- no aggregation

| value    | PR #3581 InaSAFE 4.0 | InaSAFE 3.5 |
|----------|---------|-------------|
| exposed  | 935,717 | 935,386     |
| fatalities ITB      | 387     | 387         |
| fatalities PAGER    | 123     | 124         |
| fatalities BAYES    | 772     | 979         |

Comments:
- the small number in the counts of exposed people may be explained by differences in how edges of the analysis area were handled
- there is difference of 1 with in PAGER fatality ratio - this PR rounds down estimated floating point numbers
- the number of fatalities using BAYES method has quite a difference - because this PR uses simplified model (median value for each MMI level rather than full 15999-element probability vectors)

@wonder-sk Can you provide the probability of fatality as well for both of versions?

**Comment by wonder-sk** [comment]
> @wonder-sk Can you provide the probability of fatality as well for both of versions?

@dynaryu I am not sure I understand the question... do you mean simply the probability "fatalities / exposed" from the numbers reported above? Or fatality rates for individual MMI levels? Or something else?

**Comment by dynaryu** [comment]
@wonder-sk 
The file named "Worden_berngamma_log_fat_rate_inasafe_10.csv" contains sampled fatality rate per each mmi bin. The methodology is described in the paper entitled "An Empirical Fatality Model for Indonesia Based on a Bayesian Approach", which will be published in Geohazards in Indonesia: Earth Science for Disaster Risk Reduction (https://www.geolsoc.org.uk/SP441).

The output that I'd like to check is the value of prob_fatality_mag in the code, which is the PAGER-like output. Also I wonder how each of the models perform compared to the observed fatalities for the Padang event.

**Comment by wonder-sk** [comment]
> The output that I'd like to check is the value of prob_fatality_mag in the code

I have not ported that code to InaSAFE 4 yet, so there is not much to compare there. The results from prob_fality_mag have not been used in the EQ impact report in InaSAFE 3.5, so I skipped that bit for now. If it is decided to switch the reporting of fatalities to PAGER-like output with magnitudes, I will of course reintroduce that code.

Is there any chance to get hold of the paper before it is published? I am curious how the fatality rates are sampled - are the values for different MMI levels in one row somehow related to each other or are they completely independent? (I am thinking whether we really need the whole 15999x7 matrix or whether there are ways to reduce the size of the matrix without affecting the results much).

**Comment by dynaryu** [comment]
@wonder-sk  Actually, the sampled mmi per mmi can be approximated by a known distribution, but I did not. You can find the draft version of the paper here (https://www.dropbox.com/s/bo5dylilol0djtg/SengaraEtAl_FatalityPaper_revision_August2016.pdf?dl=0).

**Comment by wonder-sk** [comment]
@dynaryu Many thanks for the link to the paper. So if I understand correctly, the values are random samples from the mixed Bernoulli/log-normal distribution as given in Eq.1 in the paper, with fixed values of parameters mu, sigma and p. Is that correct? If it is, I would be keen to update the code to draw samples from that distribution as needed rather than reading pre-generated samples every time from the existing CSV file. For a random reader of the code it would be much easier to understand what is going on. Would it be possible to share the values of mu, sigma, p used to generate the sampled values? (assuming they were fixed and not newly sampled every time from posterior distributions of a,b,c,d)

**Comment by dynaryu** [comment]
@wonder-sk 
The parameter values are not fixed but sampled as well. That's why I used the pre-sampled values.

**Comment by wonder-sk** [comment]
Just to clarify the state of implementation of the ITB Bayesian model in InaSAFE 4 as of now:
- it is not 1:1 port of InaSAFE 3.5 code - there will be some differences in the number of reported fatalities for ITB Bayesian model between InaSAFE 3.5 and 4.
- the auxiliary file with sampled probabilities (worden_berngamma_log_fat_rate_inasafe_10.csv) is not used, the implementation only uses median of each column (= one MMI zone), see: https://github.com/inasafe/inasafe/blob/develop/safe/impact_function/earthquake.py#L164
- as there is a lot of uncertainty in estimation of fatalities, it should be decided whether to keep reporting the estimated number or whether to report probability of magnitudes of fatalities (like in the image pasted above). 
- if switching to reporting of probability of magnitues, there are some caveats:
  - the "basic" ITB model does not convey information about the uncertainty, so if switching to output of probability of magnitudes, that model would need to be dropped or it would have to report 100% probability for the particular estimated magnitude.
  - the calculation of displaced people: currently it is calculated as "displaced people" = "exposed people" - "fatalities". As the number of fatalities would be a probability distribution rather than a single value, it would not make much sense to calculate that -- we would just report the total number of exposed people.

**Comment by dynaryu** [comment]
@Charlotte-Morgan @wonder-sk
I have a few comments on the changes to the ITB Bayesian model in InaSAFE 4.

1. It is proposed that the sampled fatality rate (a vector per each mmi bin) would be replaced with median fatality rate (a constant value per each mmi bin), and hard-coded. This is not a proper implementation because the result will be different. I'd like to know what concerns the most in using the file. Is it a size issue or anything else? 

2. It's possible to get other two fatality models to generate the probability of magnitudes just like the Bayesian model. Alternatively we can create fatality rate matrix for other models if the size is not an issue and if you want a seamless process regardless of fatality models.

3. For the calculation of displaced people, I would suggest use the mode for point estimate of fatality. We might need to decide the bin size (e.g., 10 or 5) for the estimation.

**Comment by Charlotte-Morgan** [comment]
@wonder-sk and @timlinux - are you able to progress this with the feedback from @dynaryu ?

**Comment by wonder-sk** [comment]
Hi @Charlotte-Morgan 

I think first we need an executive decision :-) on how to report number of fatalities - whether to 1) display estimated number or 2) display graph with probabilities of magnitudes of fatalities.

To address comments from @dynaryu:
I am aware that the ITB Bayesian model in 4.0 is not correct as mentioned above - I used the hard-coded values just to get "something" out of it when porting to 4.0. I had previously concerns about the mentioned CSV file due to unknown source and content of it - this has been mostly explained in this ticket, but it should be still documented in the code that makes use of the file. I agree using that CSV file will be easier than sampling those values on each algorithm tun. My other concern was that other EQ fatality models worked with scalars while ITB Bayesian model worked with vectors and it was not even clear in the code. Now, knowing how that model works, it would make sense to adjust the code to work with vectors for ITB Bayesian model.

Please note I currently do not have any time allocated to work on this further...

**Comment by Charlotte-Morgan** [comment]
hi @wonder-sk - I appreciate you pitching in with your advice. @dynaryu is also not funded to work on this project either so I am happy to make an executive decision that we should  retain the full functionality in the function if we can and not 'dumb it down' too much.  
For now I think we should stick with the estimated number and in 4.1+ look to change the displace to a graph.

@timlinux - over to you please to find a suitable geek to manage this into the future

**Comment by timlinux** [comment]
@Charlotte-Morgan and @wonder-sk thanks. I am proposing to remove blocker status for now, move this to 4.1 and we will add bins / histogram support in the future. @wonder-sk we have budgeted for more time from you for our new contract and so will liase with you for when you are available to work on this again if needed.

**Comment by Charlotte-Morgan** [comment]
hi @Gustry - thanks for the chat today about earthquake calcs and for raising this ticket again.
I think you have self nominated as 'suitable geek'! 
Lets explore how we can change the display to a graph as a standard way of reporting for earthquake fatalities for both InaSAFE and InaSAFE realtime.  
Adding @ivanbusthomi, @lucernae , @gubuntu and @hadighasemi back into this trail and updating the milestone.
cc @RikkiWeber , @timlinux
